### ISSUE TITLE
How can I add a custom module-path?

### ISSUE THREAD

**Issue by TomerFi** [sormuras/junit-platform-maven-plugin#54]
Hey @sormuras 

I'm sure you already know by now how much I love using this plugin.
So again, thank you very much for it!
:-)

I'm having a hard time getting a custom module-path to play nice with the plugin.
It's a private organization repository, so I'll try describing my issue first, and if need be, I'll create a demo project.

Using the [moditect plugin](https://github.com/moditect/moditect) I create a couple of modified pre-existing jars inserted with a new `module-info` descriptor.
To use the modified jars for compiling, I've added to my compiler configuration:

```xml
<properties>
  <jpms.mods.directory>${project.build.directory}/mods</jpms.mods.directory>
</properties>
<build>
  <plugin>
    <artifactId>maven-compiler-plugin</artifactId>
    <configuration>
      <compilerArgs>
        <arg>--module-path</arg>
        <arg>${jpms.mods.directory}</arg>
      </compilerArgs>
    </configuration>
  </plugin>
</build>
```

Both source and test classes are compiled successfully,
However, I can't seem the find a way to instruct this plugin to use the custom module-path as well.
It screams `java.lang.module.FindException: Module X not found, required by Y`.
X being the jar converted to modular,
Y of course being my own project.

I've tried with `NONE` and `MERGED` isolation levels,
I've tried it with both `DIRECT` and `JAVA` executors,
And I've tried setting `javaOptions.addModulesArgument` to either ALL-MODULE-PATH and ALL-DEFAULT.

I can share the log, but as far as I can see it looks the module-path constructed doesn't contain my custom one.

I might be missing something in the documentation,
Can you please point me in the right direction?

** If need be, I'll create a project for demonstration purposes.

**Comment by TomerFi** [comment]
BTW

I've tried adding the `--module-path` argument as an `additionalOption` to the `JAVA` executor, like so:

```xml
<plugin>
  <groupId>de.sormuras.junit</groupId>
  <artifactId>junit-platform-maven-plugin</artifactId>
  <extensions>true</extensions>
  <configuration>
    <isolation>MERGED</isolation>
    <executor>JAVA</executor>
    <javaOptions>
      <additionalOptions>
        <option>--module-path</option>
        <option>${jpms.mods.directory}</option>
      </additionalOptions>
    </javaOptions>
  </configuration>
</plugin>
```
The result was the same, but interesting.

This is the log from executing **without** the option set above, using the default configuration.
As expected, the default `--module-path` argument is constructed from my compiled test classes and all the jars I used in my project from my local `maven` repository:

```shell
[DEBUG] Starting process...
[DEBUG] C:\Program Files\AdoptOpenJDK\jdk-15.0.0.36-openj9\bin\java.exe
[DEBUG] ...
[DEBUG] ...
[DEBUG] -enableassertions
[DEBUG] -Dfile.encoding=UTF-8
[DEBUG] --module-path
[DEBUG] ...target/test-classes;jars-from-my-local-m2-repo...
[DEBUG] ...
[DEBUG] ...
```

After adding the above option.
The `--module-path` argument is listed twice,
One with my custom mods directory,
And one, exactly the same as before.

```shell
[DEBUG] Starting process...
[DEBUG] C:\Program Files\AdoptOpenJDK\jdk-15.0.0.36-openj9\bin\java.exe
[DEBUG] ...
[DEBUG] ...
[DEBUG] --module-path
[DEBUG] ...my-project/target/mods
[DEBUG] -enableassertions
[DEBUG] -Dfile.encoding=UTF-8
[DEBUG] --module-path
[DEBUG] ...target/test-classes;jars-from-my-local-m2-repo...
[DEBUG] ...
[DEBUG] ...
```

By the order of the lines in the log,
Is it possible that the default`--module-path` argument overrides the custom one?

**Comment by sormuras** [comment]
Short blind guess from my mobile: tweaks + additionalTestPathElements should do the trick if you're using JAVA execution mode.

**Comment by TomerFi** [comment]
I thought of that too,
But the documentation specifies it as a list of artifact coordinates.
And I need to get customize jars from build path,
I figured using the coordinates will get me the original non-modular jars from my maven repo.

```java
/** List of {@code group:artifact:version} dependencies to include in test path set. */
List<String> additionalTestDependencies = emptyList();
```

I'll give it a try and report back.

Thanks alot.

**Comment by TomerFi** [comment]
I can confirm the result is the same using:

```xml
  <tweaks>
    <additionalTestDependencies>
      <dependency>groupid:artifactid:version</dependency>
    </additionalTestDependencies>
  </tweaks>
```

**Comment by sormuras** [comment]
This one:

https://github.com/sormuras/junit-platform-maven-plugin/blob/0e39ddfbd504143519cd7423cafdb1b04d7139c6/src/main/java/de/sormuras/junit/platform/maven/plugin/Tweaks.java#L38-L39

With a usage like:

```xml
  <tweaks>
    <additionalTestPathElements>
      <element>path/to/file.jar</element>
    </additionalTestPathElements>
  </tweaks>
```

**Comment by TomerFi** [comment]
YES!
That's better.
Thank you very much.

I can now access the modules,
But unfortunately, I've now encountered  a new exception:

```shell
java.lang.module.FindException: Error reading module X
Caused by: java.lang.module.InvalidModuleDescriptorException: Package Z not found in module
```

It can now find module X.
Package Z actually comes from a missing dependency within module X,
When I initially created the fixed module-info descriptor,
I explicitly set it to ignore missing dependencies, as those dependencies are not used.
The compiler plugin seems to be ok with that, and everything works well.

Any idea on maybe making it less strict or something of that sort?

**Comment by sormuras** [comment]
Mh... 🤔 

I bet, the compiler (plugin) does run on the class-path. If it ran on the module-path, it would also fail with that error.

If the module descriptor of X says that it contains a package Z, it must contain a package Z.

Do you have package Z in another JAR file, let's call it Y? Then you may patch the content of X with Y and if there no other problems (split packages and such...), it could work.

**Comment by TomerFi** [comment]
I'm quite certain the compiler runs on the module-path.
It didn't compile prior to setting the _fixed_ module path containing the build jars.

I think I was unclear in my previous comment.
By
> I explicitly set it to ignore missing dependencies

I meant I used [--ignore-missing-deps](https://docs.oracle.com/en/java/javase/12/tools/jdeps.html#GUID-A543FEBE-908A-49BF-996C-39499367ADB4) with `jdeps` to create a manual `module-info`,
Which's content I later used to create the modular jar with [moditect](https://github.com/moditect/moditect).

I think a demo project is a must in this case,
as it involved several plugins used to "manipulate" the module-path.

I'll spin up a project and get back here.
:-)

**Comment by sormuras** [comment]
My current solution regarding those kind of issues is to strictly separate plain old JARs from Java modules. In the sense of "one bad apple spoils the bunch" - if your project (be it a library or even only an application) requires a single non-modular JAR, I stay on the class-path. 100% modules is fun, 1% less and it doesn't feel right AND steals a lot of time.

**Comment by TomerFi** [comment]
> 100% modules is fun, 1% less and it doesn't feel right AND steals a lot of time.

This is the most accurate summary of the usage of modules I've heard yet.

I totally agree with you,
I've made the same call for my previous and current big projects,
Where I had to work with some vendor's API jar.

And I probably won't merge my current JPMS migration tryout either...

But... if only they hadn't made `jlink` and `jpackage` so useful... :astonished:
I barely slept last night,
I couldn't stop thinking about how can I make this work
 :smirk:

**Comment by TomerFi** [comment]
@sormuras 
So...

I was able to spin up a [demo project ](https://github.com/TomerFi/jpms-migration-playground). 😀 

I now get a different, more clear error,
I might have misconfigured my work project.

Nevertheless, the current error I get from the plugin is:

```shell
[ERROR] java.lang.module.ResolutionException: Modules com.example.foo and foo export package com.example.foo to module org.junit.platform.commons
```

This of course is thrown by the plugin,
after the compiler successfully compiles the test classes.

I would assume the reason for that is that I have a non-modular `foo` jar on my classpath which ends up as an unnamed module on mymodulepath, and a modular `com.example.foo` jar as a named module on the same modulepath.
Both of course expose the same packages.

I need to somehow prevent `foo` from getting into the modulepath.

If you follow the README in my demo,
You can see I had a similar problem with the compiler plugin,
I was able to work around it by setting a custom `--module-path` argument.
This is of course what I initially tried to do with the junit-platform plugin, prior to opening this issue.

The original non-modular project can be found within [this commit](https://github.com/TomerFi/jpms-migration-playground/tree/07ed726f752545f5c72a0d853ebe1feb47e18320).

To sum it up:
- I have a non-modular jar called `bar`, used as the missing dependency.
- I have a non-modular jar called `foo`, depending on a local `bar` and used as the needed dependency for my project.
- I have a non-modular project called `baz`, which depends on a local `foo` but doesn't know `bar`.

My target is to convert the local `foo` within `baz` to a modular jar,
In the process, converting the whole `baz` project to a modular one.

**Comment by TomerFi** [comment]
Found It!!

Tweaks.dependencyExcludes was what I've missed.
I've used it to remove the non-modular `foo` and `additionalTestPathElements` to add the modular `foo`.

```xml
<plugin>
  <groupId>de.sormuras.junit</groupId>
  <artifactId>junit-platform-maven-plugin</artifactId>
  <version>1.1.0</version>
  <extensions>true</extensions>
  <configuration>
    <executor>JAVA</executor>
    <tweaks>
      <additionalTestPathElements>
        <element>${project.build.directory}/modules/foo-0.0.1.jar</element>
      </additionalTestPathElements>
      <dependencyExcludes>
        <exclude>com.example:foo</exclude>
      </dependencyExcludes>
    </tweaks>
  </configuration>
</plugin>
```

**Comment by TomerFi** [comment]
I Figured out my work issue as well.
It wasn't the same as the one resolved in the comment above.

```shell
java.lang.module.FindException: Error reading module com.example.foo
Caused by: java.lang.module.InvalidModuleDescriptorException: Package com.example.foo.internal.not.used not found in module
```

The issue was that the sub-package `com.example.foo.internal.not.used` from the module `com.example.foo`,
included a class that imports a class from the missing jar `bar`.

Resolving this was pretty straight-forward,
I've used the `moditect plugin` to fine-tune `foo`'s `module-info` descriptor,
so it would stop exposing the `com.example.foo.internal.not.used` subpackage using the missing `bar`.

As soon as this issue was resolved,
I of course bumped into the issue fixed by the previous comment:

```shell
java.lang.module.ResolutionException: Modules com.example.foo and foo export package com.example.foo
```

This of course the result of having `foo` in the unamed module, and `com.example.foo` as a named module.
It was resolved by using `tweaks.dependencyExcludes` as mention in the comment above.

@sormuras 
Thank you so much for this plugin,
And for your time.
:smiley:

**Comment by sormuras** [comment]
Hej Tomer,

I'm glad you figured out the underlying issues and appropriated solutions! Also, it's cool to see that the options this plugin offers are able to resolve your use-case "out-of-the-box" ... perhaps I needed those some years ago, too. ;-)

Cheers,
Christian
